modcc/procinliner.cpp

```diff
@@ -0,0 +1,208 @@
+#include "procinliner.hpp"
+
+#include <iostream>
+
+#include "astmanip.hpp"
+#include "error.hpp"
+#include "errorvisitor.hpp"
+#include "symdiff.hpp"
+
+// Note: on renaming variables when inlining:
+// Identifiers will refer to one of
+// - LOCAL variables
+// - argument,
+// - global variable
+// - PARAMETER
+// - ASSIGNED
+//
+// All local variables are renamed and the mapping is stored in local_arg_map_.
+// The mapping from arguments to local names is in call_args_map_.
+//
+// Local variable renaming of identifiers should be performed before call
+// argument renaming. This means that if a local variable shadows an argument,
+// the local variable takes precedence.
+
+void check_errors(Expression* e) {
+    ErrorVisitor v("");
+    e->accept(&v);
+    if(v.num_errors()) throw compiler_exception("something went wrong with inlined procedure call ", e->location());
+}
+
+ARB_LIBMODCC_API expression_ptr inline_procedure_calls(std::string caller, BlockExpression* block) {
+    // The inliner will inline one procedure at a time. Once all procedures in a
+    // block have been inlined, the while loop will be broken
+    auto inline_block = block->clone();
+    for(;;) {
+        inline_block->semantic(block->scope());
+
+        auto inliner = std::make_unique<ProcedureInliner>(std::move(caller));
```

Comment:
`caller` will be in _moved_ state from second iteration on - undefined behavior?

Reply:
Yes, that's a bad one. Like these do, it's copy-pasta from the function inliner.